<a href="https://colab.research.google.com/github/lavanyamurthi/Email-spam-detection-/blob/main/pincoreproject.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd


In [ ]:
data=pd.read_csv("/content/spam.tsv",sep='\t')


In [ ]:
data.head()

label                                            message  length  punct
0   ham  Go until jurong point, crazy.. Available only ...     111      9
1   ham                      Ok lar... Joking wif u oni...      29      6
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...     155      6
3   ham  U dun say so early hor... U c already then say...      49      6
4   ham  Nah I don't think he goes to usf, he lives aro...      61      2

In [ ]:
data.isnull().sum() #checking missing value

label      0
message    0
length     0
punct      0
dtype: int64

In [ ]:
data['label'].value_counts()

ham     4825
spam     747
Name: label, dtype: int64

In [ ]:
ham=data[data['label']=='ham']
ham

label                                            message  length  punct
0      ham  Go until jurong point, crazy.. Available only ...     111      9
1      ham                      Ok lar... Joking wif u oni...      29      6
3      ham  U dun say so early hor... U c already then say...      49      6
4      ham  Nah I don't think he goes to usf, he lives aro...      61      2
6      ham  Even my brother is not like to speak with me. ...      77      2
...    ...                                                ...     ...    ...
5565   ham                                       Huh y lei...      12      3
5568   ham               Will ü b going to esplanade fr home?      36      1
5569   ham  Pity, * was in mood for that. So...any other s...      57      7
5570   ham  The guy did some bitching but I acted like i'd...     125      1
5571   ham                         Rofl. Its true to its name      26      1

[4825 rows x 4 columns]

In [ ]:
spam=data[data['label']=='spam']
spam

label                                            message  length  punct
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...     155      6
5     spam  FreeMsg Hey there darling it's been 3 week's n...     147      8
8     spam  WINNER!! As a valued network customer you have...     157      6
9     spam  Had your mobile 11 months or more? U R entitle...     154      2
11    spam  SIX chances to win CASH! From 100 to 20,000 po...     136      8
...    ...                                                ...     ...    ...
5537  spam  Want explicit SEX in 30 secs? Ring 02073162414...      90      3
5540  spam  ASKED 3MOBILE IF 0870 CHATLINES INCLU IN FREE ...     158      5
5547  spam  Had your contract mobile 11 Mnths? Latest Moto...     160      8
5566  spam  REMINDER FROM O2: To get 2.50 pounds free call...     147      3
5567  spam  This is the 2nd time we have tried 2 contact u...     160      8

[747 rows x 4 columns]

In [ ]:
ham.shape , spam.shape

((4825, 4), (747, 4))

Downsampling is a process where you randomly delete some of the observations from the majority class so that the numbers in majority and minority classes are matched. Below, we have downsampled the ham messages (majority class). There are now 747 messages in each class.

In [ ]:
ham=ham.sample(spam.shape[0])
ham.shape , spam.shape

((747, 4), (747, 4))

The function x_train, x_test,y_train, y_test = train_test_split(x,y,test_size = 0.2) divides columns x and y into x_train for training inputs, y_train for training labels, x_test for testing inputs, and y_test for testing labels.

test_size=0.2 sets the testing set to 20 percent of x and y. You can see an example of this in the screenshot below, where the ham label indicates non-spam emails, and spam represents known spam emails

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(data['message'],data['label'], test_size = 0.2, random_state = 0,
                                                    shuffle = True, stratify = data['label'] )
X_train

1257       Not yet chikku..going to room nw, i'm in bus..
5461    Ok i thk i got it. Then u wan me 2 come now or...
1613    RT-KIng Pro Video Club>> Need help? info@ringt...
2179    I don,t think so. You don't need to be going o...
2638    No I don't have cancer. Moms making a big deal...
                              ...                        
3299    This message is free. Welcome to the new & imp...
1054       Jay's getting really impatient and belligerent
245                 U don't remember that old commercial?
1235    Hello-/@drivby-:0quit edrunk sorry iff pthis m...
3361                            Please attend the phone:)
Name: message, Length: 4457, dtype: object

Extracting features

fidfVectorizer - Transforms text to feature vectors that can be used as input to estimator. vocabulary_ Is a dictionary that converts each token (word) to feature index in the matrix, each unique token gets a feature index.

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
vectorizer=TfidfVectorizer()
X_train=vectorizer.fit_transform(X_train)
X_test=vectorizer.transform(X_test)

model = svm.SVC() assigns svm.SVC() to the model. In the classify.fit(features,y_train) function, model.fit trains the model with features and y_train. Then, it checks the prediction against the y_train label and adjusts its parameters until it reaches the highest possible accuracy.

In [ ]:
from  sklearn.svm import SVC
classifier=SVC(C=1000,gamma='auto')
classifier.fit(X_train,Y_train)

SVC(C=1000, gamma='auto')

**C** is a hyperparameter which is set before the training model and used to control error and **Gamma** is also a hyperparameter which is set before the training model and used to give curvature weight of the decision boundary.

In [ ]:
Y_predict=classifier.predict(X_test)

In [ ]:
from sklearn.metrics import accuracy_score,confusion_matrix
confusion_matrix(Y_test, Y_predict)

array([[962,   4],
       [ 27, 122]])

966 message were predicted as ham but out of which 4 mesg is predicted wrong indicating the 4 msg are spam
149 message were predicted as spam but out of which 27 mesg is predicted wrong indicating the 27 msg are ham

In [ ]:
print( " The accuracy level is =", accuracy_score(Y_test, Y_predict)*100, "%")

 The accuracy level is = 97.21973094170404 %


validate on the real time message

In [ ]:
classifier.predict(vectorizer.transform(['hello I lavanya' ]))

array(['ham'], dtype=object)

In [ ]:
classifier.predict(vectorizer.transform(['you won the free reword' ]))

array(['spam'], dtype=object)

In [ ]:
classifier.predict(vectorizer.transform([' Free entry in 2 a wkly comp to win FA Cup final tkts 21st may 2005' ]))

array(['spam'], dtype=object)

In [ ]:
classifier.predict(vectorizer.transform(['is that seriously how u spell his name?' ]))

array(['ham'], dtype=object)